src/openforms/tests/e2e/test_input_validation.py

```diff
@@ -670,3 +681,133 @@ def test_regex_pattern(self):
+                await page.get_by_text("blader").click()
+
+            file_chooser = await fc_info.value
+            await file_chooser.set_files(ui_files)
+            await page.wait_for_load_state("networkidle")
+
+            # try to submit the step which should be invalid, so we expect this to
+            # render the error message.
+            await page.get_by_role("button", name="Volgende").click()
+            await expect(page.get_by_text(expected_ui_error)).to_be_visible()
+
+    def test_required_field(self):
+        component = {
+            "type": "file",
+            "key": "requiredFile",
+            "label": "Required file",
+            "validate": {"required": True},
+            "storage": "url",
+        }
+
+        self.assertFileValidationIsAligned(
+            component,
+            ui_files=[],
+            expected_ui_error="Het verplichte veld Required file is niet ingevuld.",
+            api_value=[],
+        )
+
+    def test_bad_pdf(self):
+        component = {
+            "type": "file",
+            "key": "badPdf",
+            "label": "Bad PDF",
+            "validate": {"required": True},
+            "storage": "url",
+        }
+
+        # The frontend validation will *not* create a TemporaryFileUpload,
+        # as the endpoint will return a 400 because of the bad content type.
+        # For this reason, we manually create an invalid TemporaryFileUpload
+        # and use it for the `api_value`:
+        with open(TEST_FILES / "image-256x256.pdf", "rb") as infile:
+            temporary_upload = TemporaryFileUploadFactory.create(
+                file_name="image-256x256.pdf",
+                content=File(infile),
+                content_type="application/pdf",
+            )
+
+        self.assertFileValidationIsAligned(
+            component,
+            ui_files=[TEST_FILES / "image-256x256.pdf"],
+            expected_ui_error="Het bestand is geen .pdf.",
+            api_value=[
+                SubmittedFileFactory.create(
+                    temporary_upload=temporary_upload,
+                    type="application/pdf",
+                )
+            ],
+        )
+
+    def test_forbidden_file_type(self):
+        component = {
+            "type": "file",
+            "key": "badPdf",
+            "label": "Bad PDF",
+            "validate": {"required": True},
+            "storage": "url",
+            "file": {
+                "type": [
+                    "application/pdf",
+                ],
+                "allowedTypesLabels": [
+                    ".pdf",
+                ],
+            },
+            "filePattern": "application/pdf",
+        }
+
+        # The frontend validation will *not* create a TemporaryFileUpload,
+        # as the endpoint will return a 400 because of the forbidden file type.
+        # For this reason, we manually create an invalid TemporaryFileUpload
+        # and use it for the `api_value`:
```

Comment:
This comment doesn't match what is going on, but either way I would rephrase is - saying that normally the frontend blocks the upload alltogether (because of the allowed file types config), but a malicious user could be using Postman/Bruno/Curl to upload files and try to trick the backend that way.

This tests that the files are not allowed to sneak through.

I'd also add an assertion at the end that only a single temporary file upload exists to ensure no frontend regressions allow uploads after all at some point.